### PR TITLE
T0 47 be discounts ep

### PR DESCRIPTION
**Pull request by batooldshilleh**
- Implemented CRUD operations for discount management.
- Added endpoint to get remaining time for a discount.
- Added endpoint to apply discount to product price.
- Included Swagger documentation for the new endpoints.

### CODE REVIEW

File: src/controllers/productController.ts [hunk 1]
```diff
@@ -226,4 +226,32 @@ export const getProductsByBrandController = async (req: Request, res: Response):
...
+           res.status(404).json({ message: 'Product not found' });
+           return
+      }
+
+      const discount = await getDiscountByProductId(id);
+
+      if (discount) {
+          const discountAmount = product.price * (discount.discountPercentage / 100);
+          const finalPrice = product.price - discountAmount;
+           res.json({ productId: product.id, finalPrice, originalPrice: product.price });
```
Comment (Lojain-Abdalrazaq) [inline src/controllers/productController.ts:245]: finalPrice: finalPrice

**Review by TasneemJawabreh — APPROVED** [review]

File: src/controllers/productController.ts [hunk 2]
```diff
@@ -226,4 +226,32 @@ export const getProductsByBrandController = async (req: Request, res: Response):
       }
   }
 
+};
+
+export const getProductPriceAfterDiscount = async (req: Request, res: Response) => {
```
Comment (Lojain-Abdalrazaq) [inline src/controllers/productController.ts:231]: I think we have to check if the discount exists, if yes, we must make sure that it's not expired before subtracting it from the original price.

File: src/routes/productRoutes.ts [hunk 3]
```diff
@@ -36,5 +38,5 @@ router.delete("/api/products/:id", authenticateJWT, isAdmin, validateDeleteProdu
 router.get("/api/products/:id/ratings", validateGetProductRatings, getProductRatings);
 router.get("/api/products/new-arrivals", validateGetNewArrivals, getNewArrivals);
 router.get("/api/products/by-brand/:brandId", validateGetBrandProduct, validateGetAllProducts, getProductsByBrandController);
-
+router.get("/api/products/:id/price-after-discount", authenticateJWT,validateGetPriceAfterDiscount, getProductPriceAfterDiscount);
```
Comment (AlaaYa7ia) [inline src/routes/productRoutes.ts:41]: Are you sure about the structure of this rout? 

@Lojain-Abdalrazaq  what you think?

**Review by AlaaYa7ia — CHANGES_REQUESTED** [review]

File: src/routes/productRoutes.ts [hunk 4]
```diff
@@ -8,7 +8,8 @@ import {
   deleteProduct,
   getProductRatings,
   getNewArrivals,
-  getProductsByBrandController
+  getProductsByBrandController,
```
[the remark this answers is not in the document]
Comment (Lojain-Abdalrazaq) [reply src/routes/productRoutes.ts:11]: Okay, I will update this later when in my next PR. No need for Batool to do that.

File: src/routes/productRoutes.ts [hunk 5]
```diff
@@ -19,7 +20,8 @@ import {
   validateDeleteProduct,
   validateGetProductRatings,
   validateGetNewArrivals,
-  validateGetBrandProduct
+  validateGetBrandProduct,
```
[the remark this answers is not in the document]
Comment (Lojain-Abdalrazaq) [reply src/routes/productRoutes.ts:23]: Same here. @AlaaYa7ia

File: src/routes/productRoutes.ts (hunk 3 above)
Comment (Lojain-Abdalrazaq) [reply src/routes/productRoutes.ts:41]: I think its clear, and similar to the routes of the product table, for example: "/api/products/:id/ratings". 

Do you have any suggestions or ideas for improving it? @AlaaYa7ia

**Review by AlaaYa7ia — APPROVED** [review]

**Pull request merged**